src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,194 @@
+package jenkins.plugins.git;
+
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.Util;
+import hudson.model.Item;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+import org.jenkinsci.plugins.gitclient.JGitApacheTool;
+import org.jenkinsci.plugins.gitclient.JGitTool;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * A class which allows Git Plugin to choose a git implementation by estimating the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitToolChooser {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
+    /**
+     * Size to switch implementation in KiB
+     */
+    public static final int SIZE_TO_SWITCH = 5000;
+
+    /**
+     * Instantiate class using the remote name. It looks for a cached .git directory first, calculates the
+     * size if it is found else checks if the extension point has been implemented and asks for the size.
+     * @param remoteName the repository url
+     * @throws IOException
```

Comment:
Trivial additional text to silence javadoc warnings
```suggestion
     * @throws IOException on error
```